File: mocks/auth-mock/src/main/java/no/nav/pensjon/vtp/auth/azuread/MicrosoftGraphApiMock.java
package no.nav.pensjon.vtp.auth.azuread;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.annotations.Api;
import no.nav.pensjon.vtp.testmodell.ansatt.AnsatteIndeks;
import no.nav.pensjon.vtp.testmodell.ansatt.NAVAnsatt;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.toList;

@RestController
@Api(tags = {"AzureAd"})
@RequestMapping("/rest/MicrosoftGraphApi")
public class MicrosoftGraphApiMock {
    private final AnsatteIndeks ansatteIndeks;

    public MicrosoftGraphApiMock(AnsatteIndeks ansatteIndeks) {
        this.ansatteIndeks = ansatteIndeks;
    }

    @GetMapping(value= "/oidc/userinfo", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity userinfo(@RequestHeader("Authorization") String auth) {
        if (!auth.startsWith("Bearer access:")) {
            return ResponseEntity
                .status(HttpStatus.FORBIDDEN)
                    .body("Bad mock access token; must be on format Bearer access:<userid>");
        }

        String accessToken = auth.substring(14);
        String ident = accessToken.split(";")[0];

        NAVAnsatt ansatt = ansatteIndeks.hentNAVAnsatt(ident).orElseThrow(() -> new RuntimeException("Ansatt med ident " + ident + " ikke funnet i VTP."));

        Map<String, String> response = new HashMap<>();
        response.put("sub", ident);
        response.put("name", ansatt.displayName);
        response.put("family_name", ansatt.cn);
        response.put("given_name", ansatt.givenname);
        response.put("picture", "http://example.com/picture.jpg");
        response.put("email", ansatt.email);
        return ResponseEntity.ok(response);
    }

    @GetMapping(value= "/v1.0/me", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity me(@RequestHeader("Authorization") String auth, @RequestParam String select) {
        if (!auth.startsWith("Bearer access:")) {
            return ResponseEntity
                    .status(HttpStatus.FORBIDDEN)
                    .body("Bad mock access token; must be on format Bearer access:<userid>");
        }

        String accessToken = auth.substring(14);
        String ident = accessToken.split(";")[0];

        final List<Group> memberOf = ansatteIndeks.hentNAVAnsatt(ident).orElseThrow(() -> new RuntimeException("Fant ikke NAV-ansatt med brukernavn " + ident))
                .groups
                .stream()
                .map(Group::new)
                .collect(toList());

        final User user = new User("https://graph.microsoft.com/v1.0/$metadata#users(" + select + ")/$entity", ident, memberOf);

        return ResponseEntity.ok(user);
    }

    public static class User {
        @JsonProperty("@odata.context")
        public String context;
        public String onPremisesSamAccountName;

        public List<Group> memberOf;

        public User(String context, String onPremisesSamAccountName, List<Group> memberOf) {
            this.context = context;
            this.onPremisesSamAccountName = onPremisesSamAccountName;
            this.memberOf = memberOf;
        }

        @Override
        public String toString() {
            return "User{" +
                    "context='" + context + '\'' +
                    ", onPremisesSamAccountName='" + onPremisesSamAccountName + '\'' +
                    '}';
        }
    }

    public static class Group {
        public String displayName;

        public Group(String displayName) {
            this.displayName = displayName;
        }

        @Override
        public String toString() {
            return "Group{" +
                    "displayName='" + displayName + '\'' +
                    '}';
        }
    }
}

File: mocks/auth-mock/src/main/java/no/nav/pensjon/vtp/auth/azuread/MicrosoftGraphApiMock.kt
package no.nav.pensjon.vtp.auth.azuread

import com.fasterxml.jackson.annotation.JsonProperty
import io.swagger.annotations.Api
import no.nav.pensjon.vtp.testmodell.ansatt.AnsatteIndeks
import no.nav.pensjon.vtp.testmodell.ansatt.NAVAnsatt
import org.springframework.http.HttpStatus.FORBIDDEN
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import java.lang.RuntimeException

@RestController
@Api(tags = ["AzureAd"])
@RequestMapping("/rest/MicrosoftGraphApi")
class MicrosoftGraphApiMock(private val ansatteIndeks: AnsatteIndeks) {
    @GetMapping(value = ["/oidc/userinfo"], produces = [APPLICATION_JSON_VALUE])
    fun userinfo(@RequestHeader("Authorization") auth: String): UserInfo {
        val (ident, ansatt) = getAnsatt(auth)

        return UserInfo(
            sub = ident,
            name = ansatt.displayName,
            family_name = ansatt.cn,
            given_name = ansatt.givenname,
            picture = "http://example.com/picture.jpg",
            email = ansatt.email
        )
    }

    @GetMapping(value = ["/v1.0/me"], produces = [APPLICATION_JSON_VALUE])
    fun me(@RequestHeader("Authorization") auth: String, @RequestParam(required = false) select: String?): User {
        val (ident, ansatt) = getAnsatt(auth)

        return User(
            context = "https://graph.microsoft.com/v1.0/\$metadata#users($select)/\$entity",
            onPremisesSamAccountName = ident,
            memberOf = ansatt.groups.map { Group(it) }
        )
    }

    private fun getAnsatt(auth: String): Pair<String, NAVAnsatt> {
        if (!auth.startsWith("Bearer access:")) {
            throw ResponseStatusException(FORBIDDEN, "Bad mock access token; must be on format Bearer access:<userid>")
        } else {
            val ident = auth.substring(14).split(";".toRegex()).toTypedArray()[0]
            val ansatt = ansatteIndeks.findByCn(ident)
                ?: throw RuntimeException("Ansatt med ident $ident ikke funnet i VTP.")
            return Pair(ident, ansatt)
        }
    }

    data class UserInfo(
        val sub: String,
        val name: String,
        val family_name: String,
        val given_name: String,
        val picture: String,
        val email: String
    )

    data class User(
        @JsonProperty("@odata.context") var context: String,
        var onPremisesSamAccountName: String,
        var memberOf: List<Group>
    )

    data class Group(var displayName: String)
}
